File: pkg/src/paRc.c
#include "paRc.h"

#ifdef _OPENMP
#include <omp.h>
#endif

/*
 * OpenMP function wrappers
 */

void OMP_get_num_procs(int *n) {
  *n = omp_get_num_procs();
}

void OMP_set_num_threads(int *n) {
  int count = *n;
  omp_set_num_threads(count);
}

/*
 * Open Test Function
 */


void OMP_test() {
#pragma omp parallel
  {
    Rprintf("this is thread %d of %d",omp_get_thread_num(),
    omp_get_num_threads());
  }
}


void serial_calc_pi(int *n, double *pi) {
  double w = 1.0/(double) *n;
  double sum = 0, x, f_x;
  int i;
  for (i = 1; i <= *n; i++) {
    x = w*((double)i - 0.5);
    f_x = 4/(1 + x*x);
    sum += f_x;
  }
  *pi = w*sum;
}

// c't OpenMP example: calculate PI
void OMP_calc_pi(int *n, double *pi) {
  int i;
  double w, sum, x;

  w = 1.0/(double) *n;
  sum = 0;
#pragma omp parallel for private(x) shared(w) reduction(+:sum)
  for (i = 1; i <= *n; i++) {
    x = w*((double)i - 0.5);
    sum += 4.0/(1.0 + x*x);
  }
  *pi = w*sum;
}

File: pkg/R/paRc.R
## OpenMP function wrappers

## print number of available processors in the current team
omp_get_num_procs <- function(){
  out <- .C("OMP_get_num_procs", n = integer(1), PACKAGE = "paRc")
  out[[1]]
}

omp_get_max_threads <- function(){
  out <- .C("OMP_get_max_threads", n = integer(1), PACKAGE = "paRc")
  out[[1]]
}

omp_set_num_threads <- function(x){
  max_procs <- omp_get_num_procs()
  if(x < 1)
    stop("'x' must be greater than or equal 1")
  if(x > max_procs)
    warning("'x' is greater than the maximum number of available processors!")
  .C("OMP_set_num_threads", as.integer(x), PACKAGE = "paRc")
  invisible(x)
}
  

omp_test <- function(){
  .C("OMP_test", PACKAGE = "paRc")
            
}


## Approximate Pi calculation
omp_calc_pi <- function(n){
  out <- .C("OMP_calc_pi", as.integer(n), pi=double(1), PACKAGE = "paRc")
  out$pi
}

serial_calc_pi <- function(n){
  out <- .C("serial_calc_pi", as.integer(n), pi=double(1), PACKAGE = "paRc")
  out$pi
}
